src/PhpSpec/ServiceContainer/IndexedServiceContainer.php

```diff
@@ -70,61 +75,42 @@ public function getParam($id, $default = null)
      * Sets a object to be used as a service
      *
      * @param string $id
-     * @param object $value
+     * @param object $object
      * @param array  $tags
      *
-     * @throws \InvalidArgumentException if service is not an object or callable
+     * @throws \InvalidArgumentException if service is not an object
      */
-    public function set($id, $value, $tags=[])
+    public function set($id, $object, $tags=[])
     {
-        if (!is_object($value)) {
+        if (!is_object($object)) {
             throw new InvalidArgumentException(sprintf(
                 'Service should be an object, but %s given.',
-                gettype($value)
+                gettype($object)
             ));
         }
 
-        $this->addToIndex(
-            $id,
-            function() use ($value) {
-                return $value;
-            },
-            $tags
-        );
+        $this->services[$id] = $object;
+        unset($this->definitions[$id]);
+
+        $this->indexTags($id, $tags);
     }
 
     /**
-     * Sets a callable for the service creation. The same service will
+     * Sets a factory for the service creation. The same service will
      * be returned every time
      *
      * @param string   $id
-     * @param callable $callable
+     * @param callable $value
      * @param array    $tags
      *
      * @throws \InvalidArgumentException if service is not a callable
      */
-    public function define($id, $callable, $tags = [])
+    public function define($id, $value, $tags = [])
     {
-        if (!is_callable($callable)) {
-            throw new InvalidArgumentException(sprintf(
-                'Service should be callable, "%s" given.',
-                gettype($callable)
-            ));
-        }
-
-        $this->addToIndex(
-            $id,
-            function ($container) use ($callable) {
-                static $instance;
-
-                if (null === $instance) {
-                    $instance = call_user_func($callable, $container);
-                }
+        $this->definitions[$id] = $value;
+        unset($this->services[$id]);
```

Comment:
if we do describe+run, we should probably isolate these 2 calls by recreating the application, because they would not run in the same PHP process in real usage and so they cannot share state